### PR TITLE
specs: VEP proposal for VDK Properties Based Configuration

### PR DESCRIPTION
**Pull request by tozka**
This VEP proposes a new provider for VDK configuration that allows users to set job-specific configuration for data jobs.
Currently, users can only set configuration using environment variables or the config.ini file, both of which have limitations. The new provider will take configuration from properties with the "vdk.config." prefix and have lesser priority than environment variables and the config.ini file.

### CODE REVIEW

File: specs/vep-1722-properties-based-vdk-configuration-provider/README.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+- [Requirements and goals](#requirements-and-goals)
+- [High-level design](#high-level-design)
+- [API Design](#api-design)
+- [Detailed design](#detailed-design)
+- [Implementation stories](#implementation-stories)
+- [Alternatives](#alternatives)
+
+## Summary
+
+This VEP proposes a new provider for VDK configuration that allows users to set job-specific configuration for data jobs.
```
Comment (ivakoleva) [inline specs/vep-1722-properties-based-vdk-configuration-provider/README.md:39]: nit: may be useful to have a design diagram reflecting on global or specific job configurations and priority

File: specs/vep-1722-properties-based-vdk-configuration-provider/README.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+## API design
+
+The new provider for VDK configuration will not require any changes to existing API.
+
+
+
+## Detailed design
+
+
+[![sequence diagram](diagrams/sequence-diagram.svg)](https://www.plantuml.com/plantuml/uml/ZLCzZ-8m4EtzAqPGKC59b1Gd3X3IVHGqTnNIoOadYPTWM_OO_lvEecZZJKYqI4XuFjwRzsRO1vAUuhqMCYVhuMz0Bno7TAwTD0IBVwVVSFppS_4vUl7M8Rynm8yB9qSjIS7HcaANSAbZgSsujCj7BqdRL_cJ90c_RBP8PWGP86w5Q7-ll_TZmWuUwWQkfxyQMAy34dzpZmGhpc_obdD3NXlxtyaGdjQhLKEW7v8mmLkUpmV-Y3x2T4Fzp1hGg02Js6BKCW732d46X-Le0ohxbidynfHjWonHA1mFB-AMQGOkTWxvo31DxCrclUv1wV9mQEIBqjleeCdbtR8GN2K3YjbbJuaiaFAAdxAkqqsfo5BB4SLBv-PfT-2HeZT3Yg-CEpv9XekElSBGMP0auI7hYE5g2kk7Ujh_LgFegQbALohGgCHJ7qtgPc__z3XXTXQoBIPLs6wtOHiZystiFYwWhI-qqQ52DRobla_i6IlUwtS0)
```
Comment (ivakoleva) [inline specs/vep-1722-properties-based-vdk-configuration-provider/README.md:106]: I think the diagram is very good;
nit: reflect on configuration during the data job lifecycle, in favour of runtime; configs are applied at a particular point in runtime, and it's valuable to have it shown

File: specs/vep-1722-properties-based-vdk-configuration-provider/README.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
```
Comment (ivakoleva) [inline specs/vep-1722-properties-based-vdk-configuration-provider/README.md:107]: nit: users actors are all over the place; the flow is sequential, or could be -async

File: specs/vep-1722-properties-based-vdk-configuration-provider/README.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+
+## High-level design
+
+The new provider for VDK configuration will be added as a new option in the VDK configuration system.
+It will take configuration from properties with the "vdk.config." prefix and have lesser priority than environment variables and the config.ini file.
+The properties will be stored in a properties file.
+
+[![vdk properties based config plugin](diagrams/vdk-properties-based-config-plugin.svg)](https://www.plantuml.com/plantuml/uml/TP31IiD048Rl-nG_yQA7z04yIAK5qNee53qLR9fCqzLaPjcThOBuxkvQZ3NYSVz_v_k6cGQjlCQ-CoeEAjYAglHehATQhR0nTIUnGJ4JLYyTLkGFjgO2xmO0QkcTCB6YM7fnvDLImFNojZ0VnfoqZ_C5PlSfpk2AsuwEsRpI2dUoBSQVL_F5j-HavD2yb6uGbRNmphPhTbriBHUe0dAx_7hYxCOJSLvaqACih_wiYai4KYGsVjXujRhFsGH7ooILjBDlviJm3sr7Mk_7lFFC-eswM7CJlUKMcrG-7SleQGEthmAPuH1PDBfNqZm4yl2H0uR91i-oDLDAtBxx1000)
```
Comment (ivakoleva) [inline specs/vep-1722-properties-based-vdk-configuration-provider/README.md:95]: I don't understand it - what the boxes are, where do they live if those are instances, how do they relate to each over, what are the arrows and actions representing

File: specs/vep-1722-properties-based-vdk-configuration-provider/README.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+## Detailed design
+
+
+[![sequence diagram](diagrams/sequence-diagram.svg)](https://www.plantuml.com/plantuml/uml/ZLCzZ-8m4EtzAqPGKC59b1Gd3X3IVHGqTnNIoOadYPTWM_OO_lvEecZZJKYqI4XuFjwRzsRO1vAUuhqMCYVhuMz0Bno7TAwTD0IBVwVVSFppS_4vUl7M8Rynm8yB9qSjIS7HcaANSAbZgSsujCj7BqdRL_cJ90c_RBP8PWGP86w5Q7-ll_TZmWuUwWQkfxyQMAy34dzpZmGhpc_obdD3NXlxtyaGdjQhLKEW7v8mmLkUpmV-Y3x2T4Fzp1hGg02Js6BKCW732d46X-Le0ohxbidynfHjWonHA1mFB-AMQGOkTWxvo31DxCrclUv1wV9mQEIBqjleeCdbtR8GN2K3YjbbJuaiaFAAdxAkqqsfo5BB4SLBv-PfT-2HeZT3Yg-CEpv9XekElSBGMP0auI7hYE5g2kk7Ujh_LgFegQbALohGgCHJ7qtgPc__z3XXTXQoBIPLs6wtOHiZystiFYwWhI-qqQ52DRobla_i6IlUwtS0)
+
+To further elaborate on the detailed design of the new VDK plugin for configuration:
+
+1. Adding the plugin to the VDK SDK:
+   1. The new VDK plugin for configuration will be implemented as a new plugin within the VDK SDK.
+   2. For efficiency the plugin will be developed inside [vdk-plugin-control-cli](https://github.com/vmware/versatile-data-kit/tree/main/projects/vdk-plugins/vdk-plugin-control-cli) as it already has setup for accessing Properties
```
Comment (ivakoleva) [inline specs/vep-1722-properties-based-vdk-configuration-provider/README.md:112]: this is not at all reflected on the diagram, where the two   seem like a different component

**Review by ivakoleva — COMMENTED** [review]
LGTM, posting my draft review (in progress)

File: specs/vep-1722-properties-based-vdk-configuration-provider/README.md [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+## Glossary
+<!--
+Optional section which defines terms and abbreviations used in the rest of the document.
+-->
+
+## Motivation
+VDK allows users to develop and execute data jobs.
+Each data job can require its own specific configuration that needs to be set before execution.
+Currently, the only way to set configuration is using environment variables or the config.ini file.
+However, these approaches have limitations. Users cannot set environment variables for deployed jobs in the cloud, and the config.ini file is not secure for storing sensitive data.
```
Comment (ivakoleva) [inline specs/vep-1722-properties-based-vdk-configuration-provider/README.md:52]: nit: however, the config.ini file changes are tracked, and properties changes are not versioned. May elaborate on comparison

File: specs/vep-1722-properties-based-vdk-configuration-provider/README.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+To address these limitations, we propose a new provider for VDK configuration that allows users to set job-specific configuration using properties with the "vdk.config." prefix.
+This approach provides more flexibility and security in setting configuration, especially for sensitive data.
+
+The current limitations in setting job-specific configuration within VDK can also impact the system's usefulness for users.
+One issue is the lack of flexibility, as users are unable to adopt a new database plugin unless it has been pre-configured by a system administrator.
```
Comment (ivakoleva) [inline specs/vep-1722-properties-based-vdk-configuration-provider/README.md:58]: That is a very good point. The argumentation is well made.
nit: my suggestion is to include link to Github issue if available

File: specs/vep-1722-properties-based-vdk-configuration-provider/README.md (hunk 5 above)
Comment (tozka) [reply specs/vep-1722-properties-based-vdk-configuration-provider/README.md:112]: fixed this.

File: specs/vep-1722-properties-based-vdk-configuration-provider/README.md (hunk 7 above)
Comment (tozka) [reply specs/vep-1722-properties-based-vdk-configuration-provider/README.md:58]: add the issue link on the top. Though the VEP number is the issue number.

File: specs/vep-1722-properties-based-vdk-configuration-provider/README.md (hunk 6 above)
Comment (tozka) [reply specs/vep-1722-properties-based-vdk-configuration-provider/README.md:52]: It's clarified in the Alternative section.

**Review by duyguHsnHsn — APPROVED** [review]
Looks good to me. I really like the diagrams!

**Comment by tozka** [comment]
Closed due to inactivity for more than a month

**Pull request closed without merging**